doc/man/man3/Parking.3
.TH "Parking" 3 "Tue Dec 13 2022" "Gestionnaire de parkings intelligents" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Parking
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBParking\fP (\fBVec2\fP position, float minimumPrice, float startPrice, int DimX, int DimY, int id)"
.br
.RI "Constructeur parametre de la class parking\&. "
.ti -1c
.RI "vector< \fBPlace\fP > & \fBgetPlacesTab\fP ()"
.br
.ti -1c
.RI "int \fBgetNbPlaces\fP () const"
.br
.RI "Retourne le nombre de places du parking\&. "
.ti -1c
.RI "int \fBgetNbAvailablePlaces\fP () const"
.br
.ti -1c
.RI "int \fBgetMinPrice\fP () const"
.br
.ti -1c
.RI "int \fBgetStartingPrice\fP () const"
.br
.ti -1c
.RI "const double & \fBgetProfit\fP () const"
.br
.ti -1c
.RI "const int & \fBgetId\fP () const"
.br
.ti -1c
.RI "const vector< pair< double, double > > & \fBgetDataProfit\fP () const"
.br
.ti -1c
.RI "const vector< pair< double, double > > & \fBgetDataStartingPrice\fP () const"
.br
.ti -1c
.RI "const vector< pair< double, double > > & \fBgetDataNbPlaceTaken\fP () const"
.br
.ti -1c
.RI "const int & \fBgetNbTotalVisits\fP () const"
.br
.ti -1c
.RI "const \fBVec2\fP & \fBgetPos\fP () const"
.br
.RI "Donne la position (x,y) du parking - coin en haut a gauche\&. "
.ti -1c
.RI "const int & \fBgetDIMX\fP () const"
.br
.RI "Donne la largeur du parking\&. "
.ti -1c
.RI "const int & \fBgetDIMY\fP () const"
.br
.RI "Donne la hauteur du parking\&. "
.ti -1c
.RI "const double & \fBgetSuccessPourcentage\fP () const"
.br
.RI "fonction qui renvoie le pourcentage de succes\&. "
.ti -1c
.RI "bool \fBIsFull\fP ()"
.br
.RI "Fonction qui renvoie un entier, permet de savoir si le parking est plein ou non\&. "
.ti -1c
.RI "void \fBsetMinPrice\fP (float minimumPrice)"
.br
.ti -1c
.RI "void \fBsetStartingPrice\fP (float startPrice)"
.br
.ti -1c
.RI "void \fBsetNbAvailablePlaces\fP (int nbAvailablePlaces)"
.br
.ti -1c
.RI "void \fBaddUsersData\fP (\fBUtilisateur\fP unUtilisateur)"
.br
.RI "ajoute un utilisateur au tableau usersTab en vérifiant que celui-ci n'existe pas déjà dans le tableau\&. "
.ti -1c
.RI "void \fBaddToData\fP (double currentTime)"
.br
.ti -1c
.RI "void \fBincrementNbTotalVisits\fP ()"
.br
.ti -1c
.RI "void \fBincrementNbVisitsUser\fP (unsigned int id)"
.br
.RI "incrémente le nombre de visites de l'utilisateur en paramètre de 1 à chaque appel\&. "
.ti -1c
.RI "void \fBdecrementNbAvailablePlaces\fP ()"
.br
.RI "enleve une place au nombre de places dispo "
.ti -1c
.RI "void \fBincrementNbAvailablePlaces\fP ()"
.br
.RI "ajoute une place au nombre de places dispo "
.ti -1c
.RI "void \fBincrementNbAgreements\fP ()"
.br
.RI "Incrémente le nombre total de visites du parking\&. "
.ti -1c
.RI "void \fBincrementNbFinishedConv\fP ()"
.br
.RI "Incrémente de 1 le nombre de négociations effectuées par le parking\&. "
.ti -1c
.RI "void \fBupdateProfit\fP (double aPrice, float parkTime)"
.br
.RI "Ajoute au profit (du parking) la valeur\&. "
.ti -1c
.RI "void \fBupdateSuccessPercentage\fP ()"
.br
.RI "Met à jour le pourcentage de réussite des négociations du \fBParking\fP\&. "
.ti -1c
.RI "void \fBinitPlace\fP (int PcornerX, int PcornerY)"
.br
.RI "initialisation de toutes les places de parkings (positions, etc\&.\&.)\&. "
.ti -1c
.RI "bool \fBisPriceOk\fP (double price) const"
.br
.RI "vérifie qu'un prix donné est acceptable ou non, avec une marge arbitraire\&. Renvoie un booléen\&. "
.ti -1c
.RI "\fBMessage\fP \fBmanagingConversation\fP (\fBMessage\fP *aMessage) const"
.br
.RI "Le parking lit le message reçu (donné en paramètre) "
.ti -1c
.RI "\fBMessage\fP \fBconfirmConversation\fP (\fBMessage\fP *aMessage)"
.br
.RI "
.IP "\(bu" 2
Confirme la transaction ou non en fonction de la réponse de la voiture, fait les opérations en conséquence\&. 
.PP
"
.ti -1c
.RI "void \fBreconsiderPrices\fP ()"
.br
.RI "Consideration de la donnée membre successPercentage\&. "
.ti -1c
.RI "int \fBextractIntFromString\fP (string aString) const"
.br
.ti -1c
.RI "void \fBtestRegression\fP ()"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Parking::Parking (\fBVec2\fP position, float minimumPrice, float startPrice, int DimX, int DimY, int id)"

.PP
Constructeur parametre de la class parking\&. 
.PP
\fBParameters\fP
.RS 4
\fIposition\fP position du parking (position du rectangle en haut a gauche) 
.br
\fIminimumPrice\fP Prix minimum d'acceptance par minutes que le parking peut accepté -> A REVOIR 
.br
\fIstartPrice\fP Prix max que le parking propose -> A REVOIR 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "void Parking::addUsersData (\fBUtilisateur\fP unUtilisateur)"

.PP
ajoute un utilisateur au tableau usersTab en vérifiant que celui-ci n'existe pas déjà dans le tableau\&. 
.PP
\fBParameters\fP
.RS 4
\fIunUtilisateur\fP un utilisateur\&. 
.RE
.PP

.SS "\fBMessage\fP Parking::confirmConversation (\fBMessage\fP * aMessage)"

.PP

.IP "\(bu" 2
Confirme la transaction ou non en fonction de la réponse de la voiture, fait les opérations en conséquence\&. 
.PP

.IP "\(bu" 2
Cette fonction est appelée pour toute négociation, on y incrémente donc notamment le nombre de négociations
.PP
.PP
effectuées par le parking\&. 
.SS "void Parking::incrementNbAgreements ()"

.PP
Incrémente le nombre total de visites du parking\&. à l'issue des 10 dernières conversations\&. 
.SS "void Parking::incrementNbVisitsUser (unsigned int id)"

.PP
incrémente le nombre de visites de l'utilisateur en paramètre de 1 à chaque appel\&. 
.PP
\fBParameters\fP
.RS 4
\fIunUtilisateur\fP pointeur sur un utilisateur\&. 
.RE
.PP

.SS "bool Parking::isPriceOk (double price) const"

.PP
vérifie qu'un prix donné est acceptable ou non, avec une marge arbitraire\&. Renvoie un booléen\&. 
.PP
\fBParameters\fP
.RS 4
\fIprice\fP prix à vérifier 
.RE
.PP

.SS "\fBMessage\fP Parking::managingConversation (\fBMessage\fP * aMessage) const"

.PP
Le parking lit le message reçu (donné en paramètre) et en crée un nouveau (comme il s'agit forcément d'une réponse, les données du nouveau
.PP
message sont initialisées en fonction des données contenues dans le message reçu)\&.
.PP
Types de message générable :
.PP
.IP "\(bu" 2
OFFER
.IP "\(bu" 2
COUNTER_OFFER
.IP "\(bu" 2
LAST_OFFER
.IP "\(bu" 2
ACCEPT
.IP "\(bu" 2
REJECT
.PP
.PP
Types de message reçu pris en compte :
.PP
.IP "\(bu" 2
CALL
.IP "\(bu" 2
COUNTER_OFFER
.IP "\(bu" 2
LAST_OFFER
.IP "\(bu" 2
ACCEPT
.IP "\(bu" 2
REJECT 
.PP

.SS "void Parking::reconsiderPrices ()"

.PP
Consideration de la donnée membre successPercentage\&. puis modification éventuelle de minPrice et startingPrice en conséquence\&. 
.SS "void Parking::updateProfit (double aPrice, float parkTime)"

.PP
Ajoute au profit (du parking) la valeur\&. (prix de la place louée à une voiture) passée en paramètre\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Gestionnaire de parkings intelligents from the source code\&.


src/Parking.cpp
#include "Parking.h"

// CONSTRUCTEUR et DESTRUCTEUR

Parking::Parking(Vec2 position, float minimumPrice, float startPrice, int DimX, int DimY, int id)
{
    pos = (position);
    minPrice = (minimumPrice);
    startingPrice = (startPrice);
    isFull = (false);
    DIMX = DimX;
    DIMY = DimY;
    nbPlaces = ((DIMX / 2 - 1) * (DIMY / 4));
    nbAvailablePlaces = (nbPlaces);
    idP = (id);
    successPercentage = 100;
    successPercentageLastConv = 100;
    profit = 0;
    nbFinishedConv = 0;
    nbAgreement = 0;
    lastNbAgreements = 0;
    lastNbFinishedConv = 0;
    initPlace(position.x + 1, position.y + 1);
}

Parking::Parking()
{
}

Parking::~Parking()
{
}

// ACCESSEURS

vector<Place> &Parking::getPlacesTab()
{
    return placesTab;
}

int Parking::getNbPlaces() const
{
    return nbPlaces;
}

int Parking::getNbAvailablePlaces() const
{
    return nbAvailablePlaces;
}

int Parking::getMinPrice() const
{
    return minPrice;
}

int Parking::getStartingPrice() const
{
    return startingPrice;
}

const double &Parking::getProfit() const
{
    return profit;
}

const int &Parking::getDIMX() const
{
    return DIMX;
}

const int &Parking::getDIMY() const
{
    return DIMY;
}

const int &Parking::getId() const
{
    return idP;
}

bool Parking::IsFull()
{
    if (nbAvailablePlaces == 0)
    {
        isFull = true;
    }
    else
    {
        isFull = false;
    }
    return isFull;
}

const Vec2 &Parking::getPos() const
{
    return pos;
}

const double &Parking::getSuccessPourcentage() const
{
    return successPercentage;
}

// MUTATEURS
void Parking::decrementNbAvailablePlaces()
{
    nbAvailablePlaces--;
    IsFull();
}

void Parking::incrementNbAvailablePlaces()
{
    nbAvailablePlaces++;
    IsFull();
}

void Parking::setNbAvailablePlaces(int nb)
{
    nbAvailablePlaces = nb;
    IsFull();
}

void Parking::incrementNbAgreements()
{
    nbAgreement++;
    lastNbAgreements++;
}

void Parking::incrementNbFinishedConv()
{
    nbFinishedConv++;
    lastNbFinishedConv++;
}

void Parking::updateProfit(double aPrice, float parkTime)
{
    float realParkTime = (parkTime * 10);
    double nbHoursParked = (double)(realParkTime / 120);
    profit += (aPrice * nbHoursParked);
}

void Parking::updateSuccessPercentage()
{

    if (nbFinishedConv > 0)
    {
        successPercentageLastConv = lastNbAgreements * 100 / lastNbFinishedConv;
        successPercentage = nbAgreement * 100 / nbFinishedConv;
    }
    if (nbFinishedConv % 10 == 0)
    {
        lastNbAgreements = 0;
        lastNbFinishedConv = 0;
    }
}

void Parking::setMinPrice(float minimumPrice)
{
    minPrice = minimumPrice;
}

void Parking::setStartingPrice(float startPrice)
{

    startingPrice = startPrice;
}

void Parking::addUsersData(Utilisateur unUtilisateur)
{

    bool addOk = false; // on initialise un bool sur false, bool qui définit si on ajoute ou non l'utilisateur.
    if (!std::filesystem::exists("data/userData" + to_string(idP) + ".txt"))
    {
        addOk = true; // dans le cas ou le tableau est vide, pas besoin de vérifier donc on ajoute.
    }
    else if (!addOk)
    {
        int id;
        ifstream rUserData("data/userData" + to_string(idP) + ".txt");
        if (rUserData)
        {
            string line;
            while (getline(rUserData, line))
            {
                stringstream ss(line); // on crée un flux à partir de la ligne lue
                string token;          // une variable pour stocker les mots lus
                vector<string> tokens; // un tableau pour stocker les mots lus
                while (getline(ss, token, ','))
                {
                    tokens.push_back(token); // on ajoute le mot lu au tableau
                }
                id = stoi(tokens[0]);

                if (id == unUtilisateur.getId())
                {
                    addOk = false; // si on trouve le meme id deux fois, on ajoute pas.
                    break;
                }
                else
                    addOk = true;
            }
        }
        rUserData.close();
    }
    if (addOk) // si addOk est vraie en fin de procédure, on ajoute alors l'utilisateur.
    {
        ofstream userData("data/userData" + to_string(idP) + ".txt", ios::app);
        if (userData)
        {
            userData << unUtilisateur.getId() << ",";
            userData << unUtilisateur.getParkTime() << ",";
            userData << 0 << endl;
        }
        userData.close();
    }
    // usersTab.push_back(make_pair(0, unUtilisateur));
}

void Parking::addToData(double currentTime)
{
    ofstream dataProfit("data/dataProfit" + to_string(idP) + ".txt", ios::app);
    ofstream dataStartingPrice("data/dataStartingPrice" + to_string(idP) + ".txt", ios::app);
    ofstream dataNbPlaceTaken("data/dataNbPlaceTaken" + to_string(idP) + ".txt", ios::app);
    if (dataProfit)
    {
        dataProfit << currentTime << endl;
        dataProfit << profit << endl;
    }

    if (dataStartingPrice)
    {
        dataStartingPrice << currentTime << endl;
        dataStartingPrice << startingPrice << endl;
    }

    if (dataProfit)
    {
        dataNbPlaceTaken << currentTime << endl;
        dataNbPlaceTaken << (nbPlaces - nbAvailablePlaces) << endl;
    }

    dataProfit.close();
    dataNbPlaceTaken.close();
    dataStartingPrice.close();
}

void Parking::incrementNbVisitsUser(unsigned int idU)
{
    int id;
    int nbVisit;
    vector<vector<string>> linesData; // un tableau pour stocker les lignes lus
    ifstream rUserData("data/userData" + to_string(idP) + ".txt");
    if (rUserData)
    {
        string line;
        while (getline(rUserData, line))
        {
            stringstream ss(line); // on crée un flux à partir de la ligne lue
            vector<string> tokens; // un tableau pour stocker les mots lus
            string token;          // une variable pour stocker les mots lus
            while (getline(ss, token, ','))
            {
                tokens.push_back(token); // on ajoute le mot lu au tableau
            }
            id = stoi(tokens[0]);
            nbVisit = stoi(tokens[2]);
            if (id == idU)
            {
                // si on trouve le meme id deux fois, on incrémente son nombre de visite.
                nbVisit++;
                tokens[2] = to_string(nbVisit);
            }
            linesData.push_back(tokens);
        }
        rUserData.close();
    }
    ofstream userData("data/userData" + to_string(idP) + ".txt", ios::trunc);
    for (int i = 0; i < linesData.size(); i++)
    {
        if (userData)
        {
            userData << linesData[i].at(0) << ",";
            userData << linesData[i].at(1) << ",";
            userData << linesData[i].at(2) << endl;
        }
    }
    userData.close();
}

void Parking::initPlace(int PcornerX, int PcornerY)
{
    int indPl = 1;

    int nbPlCol = DIMX / 2 - 1; //(DIMX-2)*(DIMY-2) / nbPlaces;
    int nbPlLigne = DIMY / 4;   // On divise par 4 car les places font 2 de hauteur et qu'il y'en a 1 toutes les 2 lignes
    // On en repalera dans la semaine Mateo mais dcp la les places sont constuite en fonction de la taille du parking pour le remplir en entier
    // Et non pas du nombre de place
    for (int i = 0; i < nbPlLigne; i++)
    {
        for (int j = 0; j < nbPlCol; j++)
        {
            Place p = Place(Vec2(PcornerX + j * 2, PcornerY + i * 3 + i), indPl);
            placesTab.push_back(p);
            indPl++;
        }
    }
}

bool Parking::isPriceOk(double price) const
{

    bool ok = (price >= minPrice * 0.95);
    // On considère que le prix est acceptable s'il est au plus 5 % trop peu élevé.

    if (ok)
        return true;
    else
        return false;
}

Message Parking::managingConversation(Message *aMessage) const
{

    string senderString = "Car_park_" + to_string(getId());

    if (aMessage != nullptr)
    {
        string recipientString = aMessage->getSender();
        unsigned int idU = extractIntFromString(recipientString);

        double chosenPrice = -2;              // Initialisation avec une valeur arbitraire absurde
        string responseType = "INVALID_TYPE"; // Initialisation avec un type invalide

        double proposedCarPrice = aMessage->getPrice();
        string sentType = aMessage->getSubject();

        if (sentType == "CALL")
        {
            int nbVisit;
            int nbVisitU;
            vector<vector<string>> linesData; // un tableau pour stocker les lignes lues
            ifstream rUserData("data/userData" + to_string(idP) + ".txt");
            if (rUserData)
            {
                double reduc;
                string line;
                while (getline(rUserData, line))
                {
                    stringstream ss(line); // on crée un flux à partir de la ligne lue
                    vector<string> tokens; // un tableau pour stocker les mots lus
                    string token;          // une variable pour stocker les mots lus
                    while (getline(ss, token, ','))
                    {
                        tokens.push_back(token); // on ajoute le mot lu au tableau
                    }
                    int id = stoi(tokens[0]);
                    nbVisit = stoi(tokens[2]);
                    if (id == idU)
                    {
                        // si on trouve le meme id deux fois, on retient le nombre de visite de l'utilisateur.
                        nbVisitU = nbVisit;
                    }
                    linesData.push_back(tokens);
                }
                rUserData.close();
                reduc = (0.05 * nbVisitU) * startingPrice;
                chosenPrice = startingPrice - reduc;
                responseType = "OFFER";
            }
        }

        if (sentType == "COUNTER_OFFER")
        {
            int nbMessage = aMessage->getMessageNumber();
            if (nbMessage > 5)
            {
                if (proposedCarPrice < minPrice)
                {
                    chosenPrice = minPrice;
                }
                else
                {
                    chosenPrice = proposedCarPrice;
                }
                responseType = "LAST_OFFER";
            }

            else if ((proposedCarPrice < startingPrice))
            {

                /*
                    Le parking reçoit une proposition de tarif supérieur au tarif minimum qu'il accepte,
                mais il essaie quand même de faire monter le prix (en effet, dans la vie, quand on est vendeur et qu'on négocie le prix
                au marché dans les pays où il est d'usage de négocier, on essaie de faire monter le prix même
                si on a "les moyens" de vendre moins cher). La donnée membre minPrice du Parking
                n'est pas un tarif idéal pour ce dernier : l'idée est plutôt qu'on s'imagine que le parking ne peut
                pas se permettre de vendre moins cher que minPrice, sinon il ne respecte plus ses objectifs financiers.

                    On met quand même une borne supérieure startingPrice. Non pas que le parking ne veut pas vendre
                plus cher que startingPrice, mais il faut bien mettre un point de départ de la négociation
                du côté parking, tout comme il y en a un du côté Utilisateur (variable halfUserMaxPrice locale à la fonction membre managingConversation de Voiture).
                Si proposedCarPrice n'est pas suffisament proche de startingPrice, le parking fera un "pas" vers
                la Voiture en proposant un prix légèrement inférieur strictement à startingPrice (mais strictement supérieur
                à proposedCarPrice).

                */

                double deltaSup = abs(startingPrice - proposedCarPrice);
                double deltaInf = abs(proposedCarPrice - minPrice);

                if (deltaSup < deltaInf)
                {
                    // !!!!! On pourra tester avec d'autres variantes (plus avantageuses pour le parking
                    // [ou pour la voiture ?]) de cette condition.

                    chosenPrice = proposedCarPrice;

                    responseType = "ACCEPT";
                    // [SUGGGESTION :] Cela veut dire que le parking acceptera forcément la voiture S'IL A ENCORE
                    // DE LA PLACE quand la voiture arrivera à son entrée, et ce au prix chosenPrice.
                    // En revanche, cela ne veut pas dire que la voiture va aller dans ce parking (appelons-le "parking A").
                    // Ce n'est pas un accord engageant. En effet, si par la suite, dans une conversation parallèle,
                    // la voiture se met d'accord avec un autre parking pour une offre moins chère avant d'atteindre le parking A,
                    // elle n'ira pas dans le parking A.
                }

                else if (responseType != "LAST_OFFER")
                {
                    double chosenPriceTimes100 = (startingPrice - deltaSup / 2) * 100;
                    double roundedChosenPriceTimes100 = floor(chosenPriceTimes100);
                    double chosenPricePlusOneCent = roundedChosenPriceTimes100 / 100;
                    /* Les 3 lignes ci-dessus permettent d'affecter la valeur startingPrice - deltaSup / 2
                    à chosenPricePlusOneCentime, mais arrondie au centime (i.e. au centième) inférieur. On fait cela car
                    ça n'a pas de sens de proposer un prix plus précis qu'au centime près, et on arrondit
                    vers le bas car le but est que le parking diminue son prix (pour que la négociation avance !). */

                    chosenPrice = chosenPricePlusOneCent - 0.01;
                    /* On retire 1 centime au prix pour être sûr que ce dernier ne stagne pas. */

                    responseType = "COUNTER_OFFER";
                }
            }

            else
            {
                if (proposedCarPrice >= startingPrice)
                {

                    chosenPrice = proposedCarPrice;
                    responseType = "ACCEPT";
                    // [SUGGGESTION :] Cela veut dire que le parking acceptera forcément la voiture S'IL A ENCORE
                    // DE LA PLACE quand la voiture arrivera à son entrée, et ce au prix chosenPrice.
                    // En revanche, cela ne veut pas dire que la voiture va aller dans ce parking (appelons-le "parking A").
                    // Ce n'est pas un accord engageant. En effet, si par la suite, dans une conversation parallèle,
                    // la voiture se met d'accord avec un autre parking pour une offre moins chère avant d'atteindre le parking A,
                    // elle n'ira pas dans le parking A.
                }
            }
        }

        if (sentType == "LAST_OFFER")
        {
            if (isPriceOk(proposedCarPrice))
            {

                chosenPrice = proposedCarPrice;

                responseType = "ACCEPT";
                // [SUGGGESTION :] Cela veut dire que le parking acceptera forcément la voiture S'IL A ENCORE
                // DE LA PLACE quand la voiture arrivera à son entrée, et ce au prix chosenPrice.
                // En revanche, cela ne veut pas dire que la voiture va aller dans ce parking (appelons-le "parking A").
                // Ce n'est pas un accord engageant. En effet, si par la suite, dans une conversation parallèle,
                // la voiture se met d'accord avec un autre parking pour une offre moins chère avant d'atteindre le parking A,
                // elle n'ira pas dans le parking A.
            }

            else
            {
                chosenPrice = -1;
                responseType = "REJECT";
            }
        }

        if (nbAvailablePlaces == 0)
        {
            responseType = "NO_MORE_SPOTS";
            chosenPrice = -1;
        }

        if (sentType == "ACCEPT")
        {

            chosenPrice = proposedCarPrice;

            responseType = "ACCEPT";
            // [SUGGGESTION :] Cela veut dire que le parking acceptera forcément la voiture S'IL A ENCORE
            // DE LA PLACE quand la voiture arrivera à son entrée, et ce au prix chosenPrice.
            // En revanche, cela ne veut pas dire que la voiture va aller dans ce parking (appelons-le "parking A").
            // Ce n'est pas un accord engageant. En effet, si par la suite, dans une conversation parallèle,
            // la voiture se met d'accord avec un autre parking pour une offre moins chère avant d'atteindre le parking A,
            // elle n'ira pas dans le parking A.

            // TO DO : il faudra appeler une fonction qui fait que le parking stocke l'adresse de l'utilisateur
            //         pour savoir que, si ce dernier arrive à son entrée, la négociation a déjà été faite et le prix
            //         décidé.
        }

        if (sentType == "REJECT")
        {

            chosenPrice = -1;

            responseType = "REJECT";
        }

        unsigned int MessageNum = aMessage->getMessageNumber() + 1;
        Message newMessage(MessageNum, chosenPrice, responseType, senderString, recipientString);
        return newMessage;
    }

    else
    {

        string recipientString = "Unknown user";

        Message newMessage(senderString, recipientString);
        // S'il s'agit du 1er message de la conversation, la voiture ne propose
        // pas de prix (donc on met ce dernier à -1 en appelant le constructeur à 2 paramètres),
        // et le message est de type "CALL" (cf constructeur à 2 paramètres)
        // car la voiture ne fait qu'avertir le parking qu'elle veut démarrer une négociation avec lui.

        return newMessage;
    }
}

Message Parking::confirmConversation(Message *aMessage)
{
    string senderString = "Car_Park_" + to_string(getId());
    string recipientString = aMessage->getSender();
    unsigned int messageNum = aMessage->getMessageNumber() + 1;
    double price = aMessage->getPrice();
    string subject;

    if (aMessage->getSubject() == "CONFIRM_ACCEPT")
    {
        subject = "OK_TO_PARK";
    }

    else
    {
        subject = "ABORT";
        price = -1;
    }

    return Message(messageNum, price, subject, senderString, recipientString);
}

void Parking::reconsiderPrices()
{
    double nbPlDependance;
    unsigned int nbPlacesTaken = (nbPlaces - nbAvailablePlaces);
    if (nbPlacesTaken != 0)
        nbPlDependance = (((double)(nbPlacesTaken * 100 / nbPlaces) + 1) / 100);
    else
        nbPlDependance = 0;
    if (successPercentageLastConv < 50)
    {
        bool limit = false;
        bool plLimit = false;

        if (nbPlDependance > 0.85)
            plLimit = true;
        
        double reductionStartPrice = ((50 - successPercentage) / 100 * startingPrice);

        if (nbPlDependance >= 0.5 && nbPlDependance < 0.8)
            reductionStartPrice *= 0.5;
        else if (nbPlDependance >= 0.8)
            reductionStartPrice *= 0.2;
        
        if (startingPrice - reductionStartPrice >= minPrice && !plLimit)
            setStartingPrice(startingPrice - reductionStartPrice);
        else
            limit = true;

        if (limit && !plLimit)
            setMinPrice(minPrice * 0.98);
    }
    else if (successPercentageLastConv > 70)
    {
        double augmentation = 1.2 + (nbPlDependance / 2);
        if (minPrice < 7)
        {
            setMinPrice(augmentation * minPrice);
        }
        setStartingPrice(augmentation * startingPrice);
    }
}

int Parking::extractIntFromString(string aString) const
{
    const char *charString = aString.c_str(); // On convertit la string en char*
    string anIntTmp;
    int anInt;
    while (*charString)
    {
        // On avance dans la string jusqu'à trouver le premier chiffre
        if ((*charString >= '0') && (*charString <= '9'))
        {
            // On convertit le char* en int
            anIntTmp += *charString;
        }
        charString++;
    }
    anInt = stoi(anIntTmp);
    return anInt;
}

void Parking::testRegression()
{
    Parking p1(Vec2(1, 1), 0.4, 0.4, 40, 36, 1);

    // test de la focntion getPlacesTab
    assert(p1.getPlacesTab().size() == p1.placesTab.size());
    assert(p1.getPlacesTab()[0].getPos().x == p1.placesTab[0].getPos().x);
    assert(p1.getPlacesTab()[0].getPos().y == p1.placesTab[0].getPos().y);
    assert(p1.getPlacesTab()[0].getIndP() == p1.placesTab[0].getIndP());
    assert(p1.getPlacesTab()[0].getIsTaken() == p1.placesTab[0].getIsTaken());

    p1.getPlacesTab()[0].setIsTaken(false);
    cout << "Etat de la place 1 : " << p1.getPlacesTab()[0].getIsTaken() << endl;
    cout << "Etat de la place 1 : " << p1.placesTab[0].getIsTaken() << endl;

    p1.placesTab[0].setIsTaken(true);
    cout << "Etat de la place 1 : " << p1.placesTab[0].getIsTaken() << endl;
    cout << "Etat de la place 1 : " << p1.getPlacesTab()[0].getIsTaken() << endl;
    assert(p1.getPlacesTab()[0].getIsTaken() == p1.placesTab[0].getIsTaken());
    Utilisateur u1(2.5, 1, "paulo-1", " pilo", 20, 2);
    Utilisateur u2(2.5, 2, "paulo-2", " pilo", 20, 2);
    p1.addUsersData(u1);
    p1.addUsersData(u2);
    p1.addUsersData(u1);
    p1.incrementNbVisitsUser(u1.getId());
    p1.incrementNbVisitsUser(u1.getId());
}